src/libpriv/rpmostree-core.c

```diff
@@ -1526,51 +1441,45 @@ break_single_hardlink_at (int           dfd,
 
   if (stbuf.st_nlink > 1)
     {
-      g_autofree char *path_tmp = g_strconcat (path, ".XXXXXX", NULL);
-      if (!renametempat (dfd, path, dfd, path_tmp, error))
-        goto out;
+      guint count;
+      gboolean copy_success = FALSE;
+      char *path_tmp = glnx_strjoina (path, ".XXXXXX");
 
-      if (!glnx_file_copy_at (dfd, path_tmp, &stbuf, dfd, path, 0,
-                              cancellable, error))
-        goto out;
-
-      if (unlinkat (dfd, path_tmp, 0) < 0)
+      for (count = 0; count < 100; count++)
```

Comment:
I think we need 3 helper functions:

 - directories: glnx_mkdtempat
 - regfiles: glnx_mkstempat
 - symlinks: glnx_mkstemplinkat (?)

Then we'd have to split up `glnx_file_copy_at` to have `glnx_regfile_copy()` which took an existing fd, and probably hand-roll a `glnx_mkstemplinkat()` + `glnx_dfd_name_get_all_xattrs() + set_all_xattrs()` here?